src/internal/starlark/starlark.go

```diff
@@ -0,0 +1,232 @@
+package starlark
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/log"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	starjson "go.starlark.net/lib/json"
+	starmath "go.starlark.net/lib/math"
+	startime "go.starlark.net/lib/time"
+	"go.starlark.net/starlark"
+	"go.starlark.net/syntax"
+	"go.uber.org/zap"
+)
+
+const (
+	goContextKey       = "goContext"
+	scriptDirectoryKey = "scriptDirectory"
+)
+
+func init() {
+	starlark.Universe["json"] = starjson.Module
+	starlark.Universe["time"] = startime.Module
+	starlark.Universe["math"] = starmath.Module
+}
+
+// GetContext returns a context.Context that applies to in the provided Starlark thread.
+func GetContext(t *starlark.Thread) context.Context {
+	x := t.Local(goContextKey)
+	if ctx, ok := x.(context.Context); ok {
+		return ctx
+	}
+	return pctx.TODO()
+}
+
+func scriptDirectory(t *starlark.Thread) string {
+	x := t.Local(scriptDirectoryKey)
+	if s, ok := x.(string); ok {
+		return s
+	}
+	return "/dev/null"
+}
+
+func nameScript(t *starlark.Thread, abs string) string {
+	wd, err := os.Getwd()
+	if err != nil {
+		return abs
+	}
+	rel, err := filepath.Rel(wd, abs)
+	if err != nil {
+		return abs
+	}
+	return rel
+}
+
+type loadResult struct {
+	globals starlark.StringDict
+	err     error
+}
+
+func newThread(name string) *starlark.Thread {
+	return &starlark.Thread{
+		Name: name,
+		Print: func(t *starlark.Thread, msg string) {
+			log.Info(GetContext(t), msg)
+		},
+	}
+}
+
+// Options controls the Starlark execution environment.
+type Options struct {
+	// Modules are built-in modules that are available for programs to load with "load".
+	// "*.star" is always resolve to a disk path; everything else can be a key in this map.  The
+	// StringDict map value should be a map from function names to starlark.Builtin objects.
+	Modules map[string]starlark.StringDict
+	// Predefined is a dictionary of predefined values that don't need to be loaded.
+	Predefined starlark.StringDict
+	// REPLPredefined are values that are only predefined in an interactive shell.
+	REPLPredefined starlark.StringDict
+	// ThreadLocalVars are additional thread-local variables to set.  Go code can get at these
+	// environment variables with thread.Local(key).  See GetContext() for an example.
+	ThreadLocalVars map[string]any
+}
+
+// RunFn is a function that can run a Starlark program in a thread.  in is the string passed to Run,
+// module is the canonical module name for that code, if applicable.
+type RunFn func(fileOpts *syntax.FileOptions, thread *starlark.Thread, in string, module string, globals starlark.StringDict) (starlark.StringDict, error)
+
+// Run runs the provided RunFn.  The input "in" is not run here, and only serves as a base path for
+// resolving module loads.
+func Run(rctx context.Context, in string, opts Options, f RunFn) (starlark.StringDict, error) {
```

Comment:
If `cancel` is declared, `rctx` can be renamed `ctx` and `Run`’s signature is idiomatic.

Reply:
The main reason we own the context canceler is so that the REPL can fully control C-c behavior; the terminal is in raw mode and doesn't synthesize a signal from the keypress.